### PR TITLE
Add trainscript, hydra, pytorch-lightning, and wandb

### PR DESCRIPTION
**Pull request by miccio-dk**
as per title ;)

things to look out for:
- updated install/setup instructions
- added new pip + conda requirements (must install ffmpeg in docker)
- created .env file with data path
- make sure to log onto wandb beforehand (run `wandb login` -- this should also be figured out in the docker, perhaps by adding a token in the env file...)

to make sure that the setup is correct, `make train` should work.

### CODE REVIEW

File: dtu_mlops_project/configs/model/resnet34.yaml
Comment (borisrakovan) [inline dtu_mlops_project/configs/model/resnet34.yaml:1]: is this file necessary considering it's empty?

File: dtu_mlops_project/data/speechcmd.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        self.dataset_kwargs = kwargs
+
+    def setup(self, stage):
+        if stage == "fit" or stage is None:
+            self.speechcmd_train = SpeechCommands(self.data_dir, subset="training", preproc=self.train_transforms, **self.dataset_kwargs)
+            self.speechcmd_val = SpeechCommands(self.data_dir, subset="validation", preproc=self.test_transforms, **self.dataset_kwargs)
+        if stage == "test" or stage is None:
+            self.speechcmd_test = SpeechCommands(self.data_dir, subset="testing", preproc=self.test_transforms, **self.dataset_kwargs)
+
+    def train_dataloader(self):
```
Comment (borisrakovan) [inline dtu_mlops_project/data/speechcmd.py]: pls let's add type hints for the method parameters and also return values for the entire file

File: dtu_mlops_project/data/speechcmd.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+import logging
+import torch
+import torchaudio
+import pandas as pd
+from omegaconf import ListConfig
+from torch.utils.data import Dataset, DataLoader
+from lightning import LightningDataModule
+
+
+logger = logging.getLogger(__name__)
```
Comment (borisrakovan) [inline dtu_mlops_project/data/speechcmd.py]: are you using the logger anywhere?

File: dtu_mlops_project/models/model.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        self.learning_rate = learning_rate
+
+        if pretrained_weights == 'IMAGENET1K_V2':
+            self.resnet50 = torchvision.models.resnet50(weights=torchvision.models.ResNet50_Weights.IMAGENET1K_V2)
+        elif pretrained_weights == 'IMAGENET1K_V1':
+            self.resnet50 = torchvision.models.resnet50(weights=torchvision.models.ResNet50_Weights.IMAGENET1K_V1)
+        else:
+            self.resnet50 = torchvision.models.resnet50(weights=None)
+
+
```
Comment (borisrakovan) [inline dtu_mlops_project/models/model.py:46]: we don't need so many blank lines here

File: dtu_mlops_project/models/model.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import torchvision
+from lightning import LightningModule
+from torch import optim, nn
+
+
+class SpeechSpectrogramsTransferLearning(LightningModule):
```
Comment (borisrakovan) [inline dtu_mlops_project/models/model.py:10]: also this module is missing typehints everywhere

File: dtu_mlops_project/models/train_model.py [hunk 5]
```diff
@@ -1,13 +1,61 @@
...
+    # logger.info("Instantiating callbacks...")
+    # callbacks = instantiate_callbacks(cfg.get("callbacks"))
+    callbacks = []
+
+    logger.info(f"Instantiating logger <{cfg.logger._target_}>")
+    pl_logger = hydra.utils.instantiate(cfg.logger)
+
+
+    logger.info(f"Instantiating trainer <{cfg.trainer._target_}>")
+    trainer = hydra.utils.instantiate(cfg.trainer, callbacks=callbacks, logger=pl_logger)
```
Comment (borisrakovan) [inline dtu_mlops_project/models/train_model.py:33]: I maybe don't get it but does this automatically create an instance of pytorch Trainer, based simply on the config file?

File: dtu_mlops_project/configs/datamodule/speechcmd.yaml [hunk 6]
```diff
@@ -0,0 +1,38 @@
+_target_: dtu_mlops_project.data.speechcmd.SpeechCommandsDataModule
```
Comment (borisrakovan) [inline dtu_mlops_project/configs/datamodule/speechcmd.yaml]: I'd consider putting the whole `configs/` directory in the project root (one level higher), the `dtu_mlops_project` is meant to contain source code of the project and all the other static files are also outside of this directory

File: dtu_mlops_project/configs/model/resnet34.yaml
Comment (miccio-dk) [reply dtu_mlops_project/configs/model/resnet34.yaml:1]: it's a placeholder for another model

File: dtu_mlops_project/data/speechcmd.py (hunk 1 above)
Comment (miccio-dk) [reply dtu_mlops_project/data/speechcmd.py]: are these mandatory/required?

File: dtu_mlops_project/data/speechcmd.py (hunk 2 above)
Comment (miccio-dk) [reply dtu_mlops_project/data/speechcmd.py]: seemingly not, will remove

File: dtu_mlops_project/models/train_model.py (hunk 5 above)
Comment (miccio-dk) [reply dtu_mlops_project/models/train_model.py:33]: yes

File: dtu_mlops_project/configs/datamodule/speechcmd.yaml (hunk 6 above)
Comment (miccio-dk) [reply dtu_mlops_project/configs/datamodule/speechcmd.yaml]: moved

**Review by borisrakovan — APPROVED** [review]

**Review by burndjs — APPROVED** [review]
